### ISSUE TITLE
Build OSGi bundles with Gradle

### ISSUE THREAD

**Issue by bertfrees** [joeha480/dotify#106]
(*This is the discussion belonging to a PR that was originally made in another Dotify repository which is now deleted. I want to keep the discussion for reference.*)

**Comment by bertfrees** [comment]
(originally posted 25 Aug 2014 by bertfrees)

see also daisy/osgi-libs#27

**Comment by bertfrees** [comment]
(originally posted 27 Aug 2014 by bertfrees)

When building with Java 7, Gradle shows the warning "bootstrap class path not set in conjunction with -source 1.6". This can be fixed by setting `options.bootClasspath`, but that's not a very clean solution because the path is different on each system.

**Comment by bertfrees** [comment]
(originally posted 27 Aug 2014 by bertfrees)

With the new plugin the file bnd.bdn is correctly included, and
MANIFEST.MF looks O.K. now, but there are still some other problems
left.

1. The `-include: bnd.bnd` instruction is causing "split package" warnings even when that's not the case. It looks like Jar:main is included twice in the classpath.

2. The OSGI-INF/*.xml files aren't generated. I'm unsure whether this is even possible with the gradle-bundle-plugin. I've tried several solutions but non off them does the trick.
   
   One solution uses the [Felix SCR Ant task](http://felix.apache.org/documentation/subprojects/apache-felix-maven-scr-plugin/apache-felix-scr-ant-task-use.html):

   ```groovy
   configurations {
       scr
   }
   
   dependencies {
       scr group: "org.osgi", name: "org.osgi.compendium", version: "4.3.0"
       scr group: "org.apache.felix", name: "org.apache.felix.scr.ant", version: "1.13.0"
   }
   
   task scr(dependsOn: compileJava) << {
       ant.taskdef(resource: "scrtask.properties", classpath: configurations.scr.asPath)
       ant.scr(srcdir: "src", destdir: "build/classes/main")
   }
   
   jar.dependsOn scr
   ```

   The problem with this is that only Felix annotations are recognizes. There are three kinds of annotations that seem almost equivalent but I don't think they are: Felix annotations (in the `org.apache.felix.scr.annotations` package), Bnd annotations (in the `aQute.bnd.annotation.component` package) and standard annotations (in the `org.osgi.service.component.annotations` package). They are not completely interchangeable and I'm not sure if they work the same way. Bnd recognizes only Bnd annotations when generating the manifest. This is getting a bit too confusing.

   Another solution uses the [Bnd Ant task](http://www.aqute.biz/Bnd/Ant) (see also [https://github.com/tux2323/bndtools-gradle-demo](https://github.com/tux2323/bndtools-gradle-demo)):

   ```groovy
   configurations {
       bnd
   }
   
   dependencies {
       bnd group: "biz.aQute", name: "bnd", version: "1.50.0"
   }
   
   task bndprepare << {
       ant.taskdef(resource: "aQute/bnd/ant/taskdef.properties", classpath: configurations.bnd.asPath)
       ant.bndprepare(basedir: "${projectDir}", print: "false")
       def projectBuildpath = ant.properties["project.buildpath"]
       def bndBuildDependencies = files(projectBuildpath.split(':'))
       sourceSets.main.compileClasspath += bndBuildDependencies
       sourceSets.main.runtimeClasspath += bndBuildDependencies
       sourceSets.test.compileClasspath += bndBuildDependencies
       sourceSets.test.runtimeClasspath += bndBuildDependencies
   }
   
   compileJava.dependsOn bndprepare
   
   task bndbuild(dependsOn: test) << {
       ant.bnd(command: "build", exceptions: "true", basedir: "${projectDir}")
   }
   
   jar.dependsOn bndbuild
   ``` 

   The problem with this solution is that is tied to Eclipse. Bnd complaints about no "workspace" being present, which is Eclipse/bndtools terminology if I understand correctly. I had the same issue when using the Bnd API directly as explained in [http://paulonjava.blogspot.be/2013/11/building-bndtools-projects-with-gradle.html](http://paulonjava.blogspot.be/2013/11/building-bndtools-projects-with-gradle.html). I thought that bnd was supposed to be a library that could be used by several build tools but apparently this is not entirely true? I don't want to rely on an IDE to build software.

  A third solution could be the [Felix SCR Bnd plugin](http://felix.apache.org/documentation/subprojects/apache-felix-maven-scr-plugin/apache-felix-scr-bndtools-use.html). It would involve adding an instruction like `-plugin: org.apache.felix.scrplugin.bnd.SCRDescriptorBndPlugin;destdir=build/classes`. But this too is tied to Eclipse and the same issue with the three kinds of annotations holds here.

**Comment by bertfrees** [comment]
(originally posted 22 Sep 2014 by bertfrees)

Problem 2. has been fixed by updating to version 0.5 of the bundle plugin. Closing because already commited to SVN.
